File: grep.py
#!/usr/bin/env python3

import sys

def main():
	exitCode = 1
	if len(sys.argv) != 2:
		print ('Error. Please write one argument.')
		sys.exit(exitCode)

	searchString = sys.argv[1]
	lenghtSearchString = len(searchString)

	for line in sys.stdin:
		highlightArray = searchHighlightSubstring(line, searchString)

		if len(highlightArray) != 0:
			string = setColorHighlight(line, highlightArray, searchString)
			print(string, end='')
			exitCode = 0

	sys.exit(exitCode)


def getIndex(line, subString, searchPosition):
	"""Starts search subString in line from searchPosition, if subString is longer than line tail after searchPosition.

	Args:
		line(string): line for search.
		subString(string): searched string.
		searchPosition(int): index to start search from.

	Returns:
		None if line tail after searchPosition shorter than searchPosition or nothing found.
		Index(int) inside line, where found substring starts.
	"""
	lengthWithoutNewLine = len(line)-1
	validSearchLength = lengthWithoutNewLine - len(subString)
	
	if searchPosition > validSearchLength:
		return None

	for i in range(searchPosition, validSearchLength+1):
		if match(line, subString, i):
			return i

	return None


def match(line, subString, mainLineIndex):
	"""Compares subString to line from mainLineIndex.

	Args:
		line(string): line for search.
		subString(string): searched string.
		mainLineIndex(int): index from line, which needs to start search subString.

	Returns:
		True if search is successful, False otherwise.
	"""
	for i in range(len(subString)):
		if subString[i] != line[mainLineIndex + i]:
			return False
	return True
			

def searchHighlightSubstring(line, subString):
	"""Fills array with index(-es) from line, where found substring(-s) starts.

	Args:
		line(string): line for search.
		subString(string): searched string.

	Returns:
		full array, if successful search, empty if nothing found.
	"""
	arrayIndexToHighlight = []
	lenghtSubstr = len(subString)
	lengthLine = len(line)-1

	index = getIndex(line, subString, 0)
	while index != None:
		arrayIndexToHighlight.append(index)
		index = getIndex(line, subString, index + lenghtSubstr)

	return arrayIndexToHighlight


def setColorHighlight(line, arr, searchString):
	"""Returns line with highlighted found substring(-s). Adds color codes for bash.

	Args:
		line(string): line for search.
		arr(array): index(-es) inside line, where found substring(-s) starts.
		searchString(string): searched string.
	"""
	highLightedLine = ''
	position = 0
	colorHighlight = '\033[91m'
	colorDefault = '\033[0m'
	for shitIndex in arr:
		highLightedLine = highLightedLine \
						+ line[position:shitIndex] \
						+ colorHighlight \
						+ searchString \
						+ colorDefault		
		position = shitIndex + len(searchString)

	highLightedLine = highLightedLine + line[position:]
	return highLightedLine


if __name__ == "__main__":
    main()

File: grep.js
//"use strict";

if ( process.argv.length != 3 ) {
	console.log('Error. Please write one argument.');
	return false;
}

const searchString = process.argv[2];
const lenghtSearchString = process.argv[2].length;

function getIndex(line, subString, searchPosition){
	let lengthWithoutZeroIndex = line.length;
	let validSearchLength = lengthWithoutZeroIndex - subString.length;

	if ( searchPosition > validSearchLength ) { return false; }
	
	for (let i = searchPosition; i <= validSearchLength; i++) {
		if ( match(line, subString, i) ) { return i; }
	}

	return false;
};

function match(line, subString, mainLineIndex){
	for (let i = 0; i < subString.length; i++) {
		if ( subString[i] != line[mainLineIndex + i] ) { return false; }		
	}
	return true;
};

function searchHighlightSubstring(line, subString){
	let arrayIndexToHighlight = [];
	let lenghtSubstr = subString.length;
	let lengthLine = line.length-1;
	
	let index = getIndex(line, subString, 0);	
	while ( index !== false  ) {
		arrayIndexToHighlight.push(index);
		index = getIndex(line, subString, index + lenghtSubstr);
	}

	return arrayIndexToHighlight;
};

function setColorHighlight(line, arr, searchString){
	let highLightedLine = '';
	let position = 0;
	const colorHighlight = '\033[91m';
	const colorDefault = '\033[0m';

	for (var i = 0; i < arr.length; i++) {
		highLightedLine = highLightedLine + line.slice(position, arr[i]) + colorHighlight + searchString + colorDefault;
		position = arr[i] + searchString.length;
	}
	
	highLightedLine = highLightedLine + line.slice(position);
	return highLightedLine;
};

let readline = require('readline');
let rl = readline.createInterface({
  input: process.stdin,
  output: process.stdout,
  terminal: false
});

rl.on('line', function(line){
    let highlightArray = searchHighlightSubstring(line, searchString);

    if (highlightArray.length != 0) {
    	let string = setColorHighlight(line, highlightArray, searchString);
    	console.log(string);
    }
});

return true;